mlir/include/mlir/Dialect/MIOpen/MIOpen.h

```diff
@@ -0,0 +1,275 @@
+  // Only valid after the transformation has been built.
+  // The names live as long as the CoordTransformBuilder
+  void getEndNames(SmallVectorImpl<StringRef> &names);
+
+  SmallString<8> startName(uint32_t dim);
+  SmallString<8> endName(uint32_t dim);
+  uint32_t startIndex(StringRef name);
+  uint32_t endIndex(StringRef name);
+
+  int64_t startSize(StringRef name);
+  int64_t startSize(uint32_t dim);
+  int64_t endSize(StringRef name);
+  int64_t endSize(uint32_t dim);
+
+  void passThrough(StringRef name);
+  void passThrough(StringRef outName, StringRef inName);
+  void passThrough(ArrayRef<StringRef> outNames, ArrayRef<uint32_t> outDims,
+                   ArrayRef<StringRef> inNames);
+
+  // Parameters is the pre and post padding for each dimension in the order they
+  // appear as arguments. For example, padding x with 2 on the left and 1 on the
+  // right and then y with 1 above and 2 below would lead to pad({"x", "y"}, {2,
+  // 1, 1, 2})
+  void pad(ArrayRef<StringRef> names, ArrayRef<int64_t> params);
+  void pad(StringRef outName, StringRef inName, int64_t left, int64_t right);
+  void pad(ArrayRef<StringRef> outNames, ArrayRef<uint32_t> outDims,
+           ArrayRef<StringRef> inNames, ArrayRef<int64_t> params);
+
+protected:
+  CoordTransformsBuilder(mlir::Builder &builder, ArrayRef<StringRef> startNames,
+                         ArrayRef<int64_t> startShape, mlir::Location loc);
+
+  template <class T, typename = typename std::enable_if<std::is_base_of<
+                         CoordTransformsBuilder, T>::value>::type>
+  static T nextTransforms(T &previous, ArrayRef<int64_t> startShape) {
+    llvm::SmallVector<StringRef, 8> previousNames;
+    previous.getEndNames(previousNames);
+    return T(previous.b, previousNames, startShape, previous.loc);
+  }
+
+  virtual void addTransform(TransformType type, ArrayRef<int64_t> params,
+                            ArrayRef<StringRef> startNames,
+                            ArrayRef<uint32_t> startDims,
+                            ArrayRef<StringRef> endNames,
+                            ArrayRef<uint32_t> endDims) = 0;
+  virtual void extractBounds(SmallVectorImpl<int64_t> &upperBounds,
+                             SmallVectorImpl<int64_t> &lowerBounds) = 0;
+
+  virtual int64_t paddingSign() const = 0;
+
+  uint32_t nStartDims();
+  uint32_t nEndDims();
+  void defineDim(StringRef name, uint32_t dim, int64_t size);
+
+  mlir::Builder &b;
+  // It's an invariant that these go upper to lower dimensions
+  llvm::SmallVector<TransformAttr> result;
+  mlir::Location loc;
+
+private:
+  llvm::StringMap<uint32_t> startIndices;
+  llvm::SmallVector<SmallString<8>, 8> startNames;
+  llvm::SmallVector<int64_t, 8> startShape;
+
+  llvm::StringMap<uint32_t> endIndices;
+  llvm::SmallMapVector<uint32_t, SmallString<8>, 8> endNames;
+  llvm::SmallVector<int64_t, 8> endShape;
+
+  bool frozen = false;
+};
+
+class TopDownCTBuilder : public CoordTransformsBuilder {
```

Comment:
When to use `TopDownCTBuilder` and when to use `BottomUpCTBuilder` shall be described.